Refs #35838 -- WSGIRequest body handling for chunked requests

p-r-a-v-i-n opened the pull request:
#### Trac ticket number
<!-- Replace XXXXX with the corresponding Trac ticket number, or delete the line and write "N/A" if this is a trivial PR. -->

ticket-35838

#### Branch description
Fix request body handling for chunked WSGI requests without
Content-Length.

Adds tests and a minimal fix.

#### Checklist
- [x] This PR targets the `main` branch. <!-- Backports will be evaluated and done by mergers, when necessary. -->
- [x] The commit message is written in past tense, mentions the ticket number, and ends with a period.
- [x] I have checked the "Has patch" ticket flag in the Trac system.
- [x] I have added or updated relevant tests.
- [ ] I have added or updated relevant docs, including release notes if applicable.
- [ ] I have attached screenshots in both light and dark modes for any UI changes.

keranbyge commented:
Hi,

I’ve been closely following ticket #35838 (WSGIRequest body handling for chunked requests) and went through the PR, commits, and tests in detail. From my understanding, the core issue occurs when handling chunked WSGI requests without a Content-Length header, where the request body is not preserved correctly, leading to inconsistent request data consumption.

I reviewed the changes in django/core/handlers/wsgi.py, the added regression tests, and the coverage report (which currently shows 100% for the modified lines). I’m comfortable tracing the request lifecycle at the WSGI layer and debugging how Django reads and buffers request bodies for chunked inputs.

I’d like to contribute further by:

Verifying edge cases around repeated body access and middleware interaction
Reviewing test coverage for related request-handling paths
Helping refine or validate the fix against real-world WSGI server behavior

Please let me know if it’s okay for me to pick this up or explore any follow-up improvements you’d recommend. I’m happy to align with your guidance and Django’s contribution standards.

p-r-a-v-i-n commented on tests/requests_tests/tests.py:
```diff
@@ -1112,6 +1112,126 @@ def test_deepcopy(self):
+            }
+        )
+        self.assertEqual(request.body, b"")
+
```
Add these tests for recent change

codingjoe commented on django/core/handlers/wsgi.py:
```diff
@@ -71,11 +71,21 @@ def __init__(self, environ):
         self.method = environ["REQUEST_METHOD"].upper()
         # Set content_type, content_params, and encoding.
         self._set_content_type_params(environ)
+        transfer_encoding = environ.get("HTTP_TRANSFER_ENCODING", "").lower()
```
Why the `lower`?

p-r-a-v-i-n replied:
because `Transfer-Encoding` values are case-insensitive.  they could be like  `Chunked` or `CHUNKED` .

codingjoe commented on django/core/handlers/wsgi.py:
```diff
@@ -71,11 +71,21 @@ def __init__(self, environ):
         except (ValueError, TypeError):
             content_length = 0
-        self._stream = LimitedStream(self.environ["wsgi.input"], content_length)
+            if transfer_encoding and len(transfer_encoding) < 2048:
```
Why compare the length of the environment variable? What am I missing?

p-r-a-v-i-n replied:
Please correct me if I’m wrong, but I think there may be a small ambiguity here: `environ` is the WSGI environ dict, not an OS environment variable. The length check was intended as a defensive bound before parsing an unexpected header value

codingjoe replied:
Yes, we are talking about the WSGI environment. It's still a string, but I understand the boundary check now. However, I believe such a check should be happening when the environment is created, not here.

codingjoe replied:
Maybe Django should get a utility function to correctly parse comma-separated header values. Except for `Set-Cookie` most multivalue headers use the CSV style.

codingjoe reviewed (commented):
Hi,

I picked this of the random review shelf. I quickly read into the issue.

A small comment here or there might actually elevate readablity for the uninnitiated.

Quick question, how does this work with compression? I believe a chunked request needs to be fully read and then inflated, right?

Cheers!
Joe

p-r-a-v-i-n committed:
Fixed #35838 -- Preserve body for chunked requests without Content-Length
When CONTENT_LENGTH is missing, WSGIRequest limits reads from
wsgi.input to zero bytes, breaking valid HTTP/1.1 requests with
Transfer-Encoding: chunked. Detect chunked encoding and avoid
zero-length LimitedStream, preserving behavior for non-chunked
requests.

Thanks Klaas van Schelven for the report and Adam Johnson and Johannes Maron for review.

codingjoe commented on tests/requests_tests/tests.py:
```diff
@@ -1302,6 +1302,126 @@ def test_handle_raw_input_seekable_exceeds_limit_deactivated(self):
+
+    def test_long_transfer_encoding_header_ignored(self):
+        """
+        Transfer-Encoding headers longer than 2048 characters should be
```
Ah, should have seen this.

codingjoe commented on django/core/handlers/wsgi.py:
```diff
@@ -71,11 +71,21 @@ def __init__(self, environ):
         self.method = environ["REQUEST_METHOD"].upper()
         # Set content_type, content_params, and encoding.
         self._set_content_type_params(environ)
+        transfer_encoding = environ.get("HTTP_TRANSFER_ENCODING", "").lower()
         try:
             content_length = int(environ.get("CONTENT_LENGTH"))
         except (ValueError, TypeError):
             content_length = 0
-        self._stream = LimitedStream(self.environ["wsgi.input"], content_length)
+            if transfer_encoding and len(transfer_encoding) < 2048:
+                if any(
+                    enc.strip() == "chunked" for enc in transfer_encoding.split(",")
+                ):
+                    content_length = None
+
+        if content_length is None:
+            self._stream = self.environ["wsgi.input"]
+        else:
+            self._stream = LimitedStream(self.environ["wsgi.input"], content_length)
```
Might I suggest a more compacted implementation with slightly fewer branches?
```suggestion
        try:
            content_length = int(environ.get("CONTENT_LENGTH"))
        except (ValueError, TypeError):
            if (
                (transfer_encoding := environ.get("HTTP_TRANSFER_ENCODING", "").lower())
                and len(transfer_encoding) < 2048
                and any(
                    enc.strip() == "chunked" for enc in transfer_encoding.split(",")
                )
            ):
                self._stream = self.environ["wsgi.input"]
            else:
                self._stream = LimitedStream(self.environ["wsgi.input"], 0)
        else:
            self._stream = LimitedStream(self.environ["wsgi.input"], content_length)
```

codingjoe reviewed (commented):
Ok, I read a little into RFC 9112. Based on my reading, I have two questions:

1. Should we check if the content-length and transfer-encoding headers [are mutually exclusive](https://datatracker.ietf.org/doc/html/rfc9112#body.content-length)?
2. The body will be chunked by a `\r\n` with a hex chunk size directive. This doesn't seem handled by your patch, right?

p-r-a-v-i-n commented:
Hey @codingjoe !,
Sorry for late reply. 
Your feedback was a new learning for me. Thank you so much for your time to review.

> Should we check if the content-length and transfer-encoding headers [are mutually exclusive](https://datatracker.ietf.org/doc/html/rfc9112#body.content-length)?

Indeed. Thanks for raising this.

As per the RFC 9113 [Transfter encoding should get precedence if both content-length and tranfter encoding present](https://datatracker.ietf.org/doc/html/rfc9112#section-6.3.3:~:text=If%20a%20message%20is%20received%20with%20both%20a%20Transfer%2DEncoding%20and%20a%20Content%2DLength%20header%20field%2C%20the%20Transfer%2DEncoding%20overrides%20the%20Content%2DLength%2E) , in such situation how should we handle this.
1. should we raise error ( [as suggested](https://datatracker.ietf.org/doc/html/rfc9112#section-6.3.3:~:text=ought%20to%20be%20handled%20as%20an%20error)  ) ? or 
2. should we removed content-length and process only transfter-encoding ?

codingjoe commented:
> 1. should we raise error ( [as suggested](https://datatracker.ietf.org/doc/html/rfc9112#section-6.3.3:~:text=ought%20to%20be%20handled%20as%20an%20error)  ) ? or
> 2. should we removed content-length and process only transfter-encoding ?

I read this too and believe this should result in a 400 response.

While the oversized header should lead to a  `[431](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Status/431)` response. But again, I don't believe you need to handle this.

p-r-a-v-i-n commented on django/core/handlers/wsgi.py:
```diff
@@ -71,11 +72,32 @@ def __init__(self, environ):
+        # ought to be handled as an error.
+        if transfer_encoding and environ.get("CONTENT_LENGTH") is not None:
+            raise BadRequest(
+                "Request cannot have both Content-Length and "
+                "Transfer-Encoding headers."
```
I think this message could be enhanced. I would appreciate any suggestions
